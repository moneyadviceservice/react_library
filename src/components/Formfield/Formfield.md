### The `Formfield` component is used to group several controls within the `Form`.
Does your form contain multiple sections of related inputs? Use `Formfield` to generate a `fieldset` element to group them, and the `legend` prop to provide a label for what this section is for.

Screen readers may repeat the legend for each control in the group, so the legend text should be brief and descriptive.
#### **Note:**
Fieldset and legend should only be used when a higher-level label is necessary. Single checkboxes or basic radio buttons that make sense from their labels alone do not require fieldset and legend.<br/>
Nested fieldsets can cause odd screen reader behavior and should be avoided.
```jsx
import { Form, Formfield, Radio, Button } from '@moneypensionservice/directories';

<Form>
  <Formfield 
    legend="Choose your favorite monster:"
    margin={{bottom: "1rem"}} >
    <Radio
      value="kraken"
      label="Kraken" />
    <Radio
      value="sasquatch"
      label="Sasquatch" />
    <Radio
      value="mothman"
      label="Mothman" />
  </Formfield>
  <Formfield legend="Choose your favorite food:">
    <Radio
      value="pizza"
      label="Pizza" />
    <Radio
      value="burger"
      label="Burger" />
    <Radio
      value="steak"
      label="Steak" />
  </Formfield>
</Form>
```

