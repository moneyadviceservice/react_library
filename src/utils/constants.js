export const breakpoints = {
  xs: 1,
  sm: 48, // 768px
  md: 64, // 1024px
  lg: 90, // 1440px
  xl: 120, // 1920px
}

export const dimensions = ['xs', 'sm', 'md', 'lg', 'xl']

export const mediaQuery = 'only screen'

export const ALIGN_SELF_MAP = {
  center: 'center',
  end: 'flex-end',
  start: 'flex-start',
  stretch: 'stretch',
}

export const SPACING_SIZES = [
  'xxsmall',
  'xsmall',
  'small',
  'medium',
  'large',
  'xlarge',
]

