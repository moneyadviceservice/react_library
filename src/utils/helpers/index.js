export * from './background'
export * from './responsive'
export * from './flexbox'
export * from './styles'
export * from './mediaQueries'
export * from './sizing'
export * from './refs'
