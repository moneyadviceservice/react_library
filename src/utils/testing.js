import { render, act } from '@testing-library/react'
import GlobalTheme from '../components/ThemeProvider'

const customRender = (ui, options) =>
  render(ui, { wrapper: GlobalTheme, ...options })

const waitRender = async (ui, options) => {
  const { container, ...rest } = customRender(ui, options)

  await act(async () => {
    await Promise.resolve(container)
  })

  return { container, ...rest }
}

export * from '@testing-library/react'

export { customRender as render, waitRender }
