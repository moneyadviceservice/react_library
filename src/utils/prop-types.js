import PropTypes from 'prop-types'
import { SPACING_SIZES } from './constants'

/** spacing prop.type */
const spacingProp = PropTypes.shape({
  bottom: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  horizontal: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  left: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  right: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
  top: PropTypes.oneOfType([PropTypes.oneOf(SPACING_SIZES), PropTypes.string]),
  vertical: PropTypes.oneOfType([
    PropTypes.oneOf(SPACING_SIZES),
    PropTypes.string,
  ]),
})

export const genericPropTypes = {
  a11yTitle: PropTypes.string,
  alignSelf: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  margin: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  padding: PropTypes.oneOfType([
    PropTypes.oneOf([...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
  border: PropTypes.oneOfType([
    PropTypes.oneOf(['none', ...SPACING_SIZES]),
    spacingProp,
    PropTypes.string,
    PropTypes.object,
  ]),
}

export function genericPropsDefaults({
  a11yTitle = null,
  alignSelf = null,
  margin = null,
  padding = null,
  border = null,
} = {}) {
  return {
    a11yTitle: a11yTitle,
    alignSelf: alignSelf,
    margin: margin,
    padding: padding,
    border: border,
  }
}
