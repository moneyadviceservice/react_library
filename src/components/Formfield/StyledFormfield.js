import styled from 'styled-components'
import { genericStyles } from '../../utils/helpers'

const StyledFormfield = styled.fieldset`
  ${genericStyles}
  ${props => !props.padding && `padding: 0 1.125rem;`}
`

const Legend = styled.legend`
  font-size: 1.125rem;
  line-height: 1.5rem;
  font-weight: 700;
  color: ${({ theme, color }) =>
    color ? color : theme.colors.formfield.legendText};
  width: 100%;
  margin-bottom: 0.75rem;
`

export { StyledFormfield, Legend }
