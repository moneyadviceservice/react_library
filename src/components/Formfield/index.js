import React from 'react'
import PropTypes from 'prop-types'
import { genericPropTypes, genericPropsDefaults } from '../../utils/prop-types'
import { StyledFormfield, Legend } from './StyledFormfield'

const Formfield = ({ a11yTitle, children, legend, legendColor, ...rest }) => {
  return (
    <StyledFormfield aria-label={a11yTitle || legend} {...rest}>
      {legend && <Legend color={legendColor}>{legend}</Legend>}
      {children}
    </StyledFormfield>
  )
}

// Documentation
Formfield.propTypes = {
  /** Content inside component. */
  children: PropTypes.node,
  /** Creates a legend element that represents a caption for the content. */
  legend: PropTypes.string,
  /** Changes the color of the legend text. */
  legendColor: PropTypes.string,
  ...genericPropTypes,
}

Formfield.defaultProps = {
  ...genericPropsDefaults(),
}

/** @component */
export { Formfield }
