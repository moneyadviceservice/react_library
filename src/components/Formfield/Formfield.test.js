import React from 'react'
import { render } from '../../utils/testing'
import { Formfield } from '../Formfield'

describe('<Formfield />', () => {
  describe('Snapshots', () => {
    it('should match snapshot with legend prop and children element', () => {
      const { container } = render(
        <Formfield legend="example">
          <p>example child</p>
        </Formfield>
      )
      expect(container).toMatchSnapshot()
    })
  })
})
