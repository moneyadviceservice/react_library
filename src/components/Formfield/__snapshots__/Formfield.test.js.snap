// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Formfield /> Snapshots should match snapshot with legend prop and children element 1`] = `
<div>
  <fieldset
    aria-label="example"
    class="StyledFormfield-yeky1l-0 kIEUKv"
  >
    <legend
      class="StyledFormfield__Legend-yeky1l-1 bhMqHd"
    >
      example
    </legend>
    <p>
      example child
    </p>
  </fieldset>
</div>
`;
